src/services/codefixes/convertToTypeOnlyImport.ts

```diff
@@ -1,67 +1,69 @@
-            const importDeclaration = getImportDeclarationForDiagnosticSpan(diag, context.sourceFile);
-            fixSingleImportDeclaration(changes, importDeclaration, context);
+            const declaration = getDeclaration(diag.file, diag.start);
+            if (declaration) {
+                doChange(changes, diag.file, declaration);
+            }
         });
     }
 });
 
-function getImportDeclarationForDiagnosticSpan(span: TextSpan, sourceFile: SourceFile) {
-    return tryCast(getTokenAtPosition(sourceFile, span.start).parent, isImportDeclaration);
+function getDeclaration(sourceFile: SourceFile, pos: number) {
+    const { parent } = getTokenAtPosition(sourceFile, pos);
+    return isImportSpecifier(parent) || isImportDeclaration(parent) && parent.importClause ? parent : undefined;
 }
 
-function fixSingleImportDeclaration(changes: textChanges.ChangeTracker, importDeclaration: ImportDeclaration | undefined, context: CodeFixContextBase) {
-    if (!importDeclaration?.importClause) {
-        return;
+function doChange(changes: textChanges.ChangeTracker, sourceFile: SourceFile, declaration: ImportDeclaration | ImportSpecifier) {
+    if (isImportSpecifier(declaration)) {
+        changes.replaceNode(sourceFile, declaration, factory.updateImportSpecifier(declaration, /*isTypeOnly*/ true, declaration.propertyName, declaration.name));
     }
-
-    const { importClause } = importDeclaration;
-    // `changes.insertModifierBefore` produces a range that might overlap further changes
-    changes.insertText(context.sourceFile, importDeclaration.getStart() + "import".length, " type");
-
-    // `import type foo, { Bar }` is not allowed, so move `foo` to new declaration
-    if (importClause.name && importClause.namedBindings) {
-        changes.deleteNodeRangeExcludingEnd(context.sourceFile, importClause.name, importDeclaration.importClause.namedBindings);
-        changes.insertNodeBefore(context.sourceFile, importDeclaration, factory.updateImportDeclaration(
-            importDeclaration,
-            /*modifiers*/ undefined,
-            factory.createImportClause(
-                /*isTypeOnly*/ true,
-                importClause.name,
-                /*namedBindings*/ undefined),
-            importDeclaration.moduleSpecifier,
-            /*assertClause*/ undefined));
+    else {
+        const importClause = declaration.importClause as ImportClause;
+        if (importClause.name && importClause.namedBindings) {
+            changes.replaceNodeWithNodes(sourceFile, declaration, [
+                factory.createImportDeclaration(declaration.modifiers, factory.createImportClause(/*isTypeOnly*/ true, importClause.name, /*namedBindings*/ undefined), declaration.moduleSpecifier, declaration.assertClause),
+                factory.createImportDeclaration(declaration.modifiers, factory.createImportClause(/*isTypeOnly*/ true, /*name*/ undefined, importClause.namedBindings), declaration.moduleSpecifier, declaration.assertClause),

```

Comment:
Is it safe to reuse `declaration.moduleSpecifier` and `declaration.assertClause`? I believe our refactoring infrastructure has issues if you share nodes while formatting. @andrewbranch might know more.

Reply:
(Maybe add a test for this?)

Reply:
I don’t 100% know exactly when/why problems occur, but I would say to be safe, if you’re duplicating a Node, it’s best to clone it. I suspect this is fine, but it’s hard to say.

Reply:
I suspect it'd be fine since this just goes into textChanges... but I can't say for sure. A test would be nice (unless I've missed an explicit test for this particular code; I'm a little confused by the array).

Reply:
The emitter will sometimes try to look at original positions of a node to figure out how many line breaks to include, so it's possible something in that could get weird.

Reply:
Well, it's more that when we format the nodes, we will have some synthetic `pos`/`end` onto nodes.

Then it typically triggers a debug assert in `consumeTokenAndAdvanceScanner`/`processChildNodes`.

https://github.com/microsoft/TypeScript/blob/2a8436c529640327a3460463fc945694584ff1c3/src/services/formatting/formatting.ts#L968-L969

See #44301 and #44523.